# Exercise 6

1. **Execute o código abaixo em um arquivo cpp da seguinte maneira**:
    
    * Na pasta ompenmp/notebooks/codigo_externo, crie um aquivo chamado: 
        * "exercise_6.cpp".

    * Clique duas vezes para abrí-lo no editor do VSCode.

    * Copie e cole o código abaixo dentro do arquivo e salve.

    * No terminal Linux, vá até a pasta do aquivo: 
        * cd ompenmp/notebooks/codigo_externo

    * Compile o aquivo:
        * g++ -fopenmp exercise_6.cpp -o exercise_6.exe

    * Execute o programa:
        ./exercise_6.exe

2. **Execute o código no notebook (single thread) e depois fora do notebook (número de threads == número de núcleos).**
    * Insira o código necessário para iniciar a contagem do tempo antes do paralelismo
    * Insira o código necessário para calcular o tempo corrigo após o término do paralelismo
    * Faça 3 execuções do código fora do notebook e, para cada uma, faça:
        * **Varie o valor de "n" e o número de threads**    
        * Imprima os resultados no console
        * Copie o tempo para a célula do tipo markdown que se encontra abaixo da célula de código
    * Compare e comente a variação nos tempos.

3. **Volte a este notebook**:
    * Na célula do tipo markdown abaixo da célula que contém o código a ser executado.

**Obs: dentro do notebook é executada apenas uma thread. Por isso, execute fora do notebook para ver o paralelismo.** 


**O código abaixo apresenta bugs.**

**Tente determinar o que está causando o erro e corrija.**

**Em seguida, execute conforme as instruções acima.**

In [11]:
#include "notebooks_reserved_code/openmp_config.h"
#include <omp.h>
#include <stdio.h>
#include <stdlib.h>

int main () 
{
	int nthreads, i, tid;
	float total = 0.0;
	int n = 100000000;
	double start_time, end_time;

	start_time = omp_get_wtime();
	/*** Spawn parallel region ***/
	#pragma omp parallel private(tid) shared(nthreads) reduction(+:total)
	{
		/* Obtain thread number */
		tid = omp_get_thread_num();
		/* Only master thread does this */
		if (tid == 0){
			nthreads = omp_get_num_threads();
			printf("Number of threads = %d\n", nthreads);
		}

		printf("Thread %d is starting...\n",tid);

		#pragma omp barrier
		/* do some work */
		
		#pragma omp for schedule(dynamic,10)
		for (i=0; i<n; i++){
			total += i * 1.0;
		}

		printf ("Thread %d is done! \n", tid);

	  } /*** End of parallel region ***/
	  end_time = omp_get_wtime();

	  printf("Time taken = %f seconds\n", end_time - start_time);
	  printf("Total = %f\n", total);


}
main()


Number of threads = 1
Thread 0 is starting...
Thread 0 is done! 
Time taken = 0.505183 seconds
Total = 2251799813685248.000000


0

## Resultados:

Cole os resultados das execuções.

| Valor de `n` | Tempo (single-thread) | Tempo (multi-thread) |
|--------------|-----------------------|----------------------|
| 1.000.000    | 0.006284              | 0.006290             | 
| 10.000.000   | 0.050105              | 0.026590             |
| 100.000.000  | 0.505183              | 0.199909             |

Com n = 1.000.000, o tempo com uma e várias threads foi praticamente o mesmo, indicando que o overhead do paralelismo compensou o ganho. Já com n = 10.000.000, houve melhora perceptível no desempenho, com quase o dobro de velocidade. Para n = 100.000.000, o ganho foi ainda maior, com tempo reduzido em mais da metade, mostrando que o uso de OpenMP se torna eficiente à medida que a carga de trabalho aumenta.


#### n = 1000000 (Single-thread)
![n=1000000 single-thread](images/ex6/single1.png)

---

#### n = 1000000 (Multi-thread)
![n=1000000 multi-thread](images/ex6/multi1.png)

---

#### n = 10000000 (Single-thread)
![n=10000000 single-thread](images/ex6/single2.png)

---

#### n = 10000000 (Multi-thread)
![n=10000000 multi-thread](images/ex6/multi2.png)

---

#### n = 100000000 (Single-thread)
![n=100000000 single-thread](images/ex6/single3.png)

---

#### n = 100000000 (Multi-thread)
![n=100000000 multi-thread](images/ex6/multi3.png)


Cole também o código corrigido.

O código apresenta condição de corrida, porque várias threads estão acessando e modificando a variável total ao mesmo tempo, sem controle de concorrência. Além disso, cada thread está zerando total dentro da região paralela, sobrescrevendo o valor das outras. Isso gera um resultado final incorreto. 


```c
#include "openmp_config.h"
#include <omp.h>
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>
#include <stdio.h>
#include <stdlib.h>

int main () 
{
	int nthreads, i, tid;
	float total = 0.0;
	double start_time, end_time;
    int n = 100000000;
	start_time = omp_get_wtime();
	/*** Spawn parallel region ***/
	#pragma omp parallel private(tid) shared(nthreads) reduction(+:total)
	{
		/* Obtain thread number */
		tid = omp_get_thread_num();
		/* Only master thread does this */
		if (tid == 0){
			nthreads = omp_get_num_threads();
			printf("Number of threads = %d\n", nthreads);
		}

		printf("Thread %d is starting...\n",tid);

		#pragma omp barrier
		/* do some work */
		
		#pragma omp for schedule(dynamic,10)
		
		for (i=0; i<n; i++){
			total += i * 1.0;
		}

		printf ("Thread %d is done! \n", tid);

	  } /*** End of parallel region ***/
	  end_time = omp_get_wtime();
	  printf("Time taken = %f seconds\n", end_time - start_time);
	  printf("Total = %f\n", total);


	  return 0;

}
main()

```

## Entrega:

**Salve este notebook com as suas respostas e poste como entrega da atividade no Canvas.**